### PR TITLE
Use ChainRules.jl

### PR DESCRIPTION
**Pull request by mcabbott**
This replaces ZygoteRules.jl with ChainRulesCore.jl in the most direct way.

Gradients are still all calculated at once, and hence cannot be independently `@thunk`-ed.

### CODE REVIEW

File: test/parsing.jl [hunk 1]
```diff
@@ -494,7 +494,7 @@ end
 
     # initialisation
     @test 200 == @tullio (max) m := A[i] init=200
-    @tullio (max) C[i] := i^2   (i in 1:10, j in 1:1)  init=33.3 # widens type
+    @tullio (max) C[i] := i^2   (i in 1:10, j in 1:1)  init=33.3 avx=false # widens type
```
Comment (chriselrod) [inline test/parsing.jl:497]: Did this throw an error or get the incorrect answer? It seems to work for me.
```julia
julia> @tullio (max) C[i] := i^2   (i in 1:10, j in 1:1)  init=33.3
10-element Vector{Float64}:
  33.3
  33.3
  33.3
  33.3
  33.3
  36.0
  49.0
  64.0
  81.0
 100.0

julia> @tullio (max) C[i] := i^2   (i in 1:10, j in 1:1)  init=33.3 avx=false
10-element Vector{Float64}:
  33.3
  33.3
  33.3
  33.3
  33.3
  36.0
  49.0
  64.0
  81.0
 100.0
```

File: test/parsing.jl [hunk 2]
```diff
@@ -494,7 +494,7 @@ end
...
     @test C == max.(33.3, A)
     @tullio C[i] := 0   (i in 1:10, j in 1:1)  init=randn() avx=false tensor=false # UndefVarError: ##op#1444_0__1 not defined
```
Comment (chriselrod) [inline test/parsing.jl:499]: https://github.com/JuliaSIMD/LoopVectorization.jl/commit/3fa854d8be285e33d20b57c4aa4ab527382355a1

**Review by chriselrod — COMMENTED** [review]
Not a real review, I'm just procrastinating.

File: test/parsing.jl (hunk 2 above)
Comment (chriselrod) [reply test/parsing.jl:499]: After the above commit:
```julia
julia> @tullio C[i] := 0   (i in 1:10, j in 1:1)  init=randn() tensor=false
10-element Vector{Float64}:
 0.1812303623999893
 0.1812303623999893
 0.1812303623999893
 0.1812303623999893
 0.1812303623999893
 0.1812303623999893
 0.1812303623999893
 0.1812303623999893
 0.1812303623999893
 0.1812303623999893
```

File: test/parsing.jl (hunk 1 above)
Comment (mcabbott) [reply test/parsing.jl:497]: I didn't investigate further than seeing an error in CI, here:

https://github.com/mcabbott/Tullio.jl/runs/2921997794#step:6:301

But locally it seems to work fine for me, too!

**Pull request merged**